From ce953f011356257c1e1862d0353f785453559028 Mon Sep 17 00:00:00 2001
From: Dyami Caliri <dyami@dragonframe.com>
Date: Tue, 16 Sep 2014 17:32:22 -0700
Subject: [PATCH] OS X: Handle Qt::AA_MacDontSwapCtrlAndMeta for regular key
 events

The flag Qt::AA_MacDontSwapCtrlAndMeta was only being used with shortcut
detection. Now trying to incorporate flag in normal key events.

Task-number: QTBUG-22775
Change-Id: I9edf97d0de93aa225435c9802372e33641cdac73
---
 src/plugins/platforms/cocoa/qnsview.mm | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/qtbase/src/plugins/platforms/cocoa/qnsview.mm b/qtbase/src/plugins/platforms/cocoa/qnsview.mm
index cfcbb80..7b65fad 100644
--- a/qtbase/src/plugins/platforms/cocoa/qnsview.mm
+++ b/qtbase/src/plugins/platforms/cocoa/qnsview.mm
@@ -758,7 +758,8 @@ static NSString *_q_NSWindowDidChangeOcclusionStateNotification = nil;
             [inputManager handleMouseEvent:theEvent];
         }
     } else {
-        if ([QNSView convertKeyModifiers:[theEvent modifierFlags]] & Qt::MetaModifier) {
+        const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
+        if ([QNSView convertKeyModifiers:[theEvent modifierFlags]] & (swapCtrlAndMeta ? Qt::MetaModifier : Qt::ControlModifier)) {
             m_buttons |= Qt::RightButton;
             m_sendUpAsRightButton = true;
         } else {
@@ -1356,15 +1357,16 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
 
 + (Qt::KeyboardModifiers) convertKeyModifiers : (ulong)modifierFlags
 {
+    const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
     Qt::KeyboardModifiers qtMods =Qt::NoModifier;
     if (modifierFlags &  NSShiftKeyMask)
         qtMods |= Qt::ShiftModifier;
     if (modifierFlags & NSControlKeyMask)
-        qtMods |= Qt::MetaModifier;
+        qtMods |= (swapCtrlAndMeta ? Qt::MetaModifier : Qt::ControlModifier);
     if (modifierFlags & NSAlternateKeyMask)
         qtMods |= Qt::AltModifier;
     if (modifierFlags & NSCommandKeyMask)
-        qtMods |= Qt::ControlModifier;
+        qtMods |= (swapCtrlAndMeta ? Qt::ControlModifier : Qt::MetaModifier);
     if (modifierFlags & NSNumericPadKeyMask)
         qtMods |= Qt::KeypadModifier;
     return qtMods;
@@ -1388,10 +1390,11 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     quint32 nativeScanCode = 1;
     quint32 nativeVirtualKey = [nsevent keyCode];
 
+    const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
     QChar ch = QChar::ReplacementCharacter;
     int keyCode = Qt::Key_unknown;
     if ([characters length] != 0) {
-        if (((modifiers & Qt::MetaModifier) || (modifiers & Qt::AltModifier)) && ([charactersIgnoringModifiers length] != 0))
+        if (((modifiers & (swapCtrlAndMeta ? Qt::MetaModifier : Qt::ControlModifier)) || (modifiers & Qt::AltModifier)) && ([charactersIgnoringModifiers length] != 0))
             ch = QChar([charactersIgnoringModifiers characterAtIndex:0]);
         else
             ch = QChar([characters characterAtIndex:0]);
@@ -1476,6 +1479,7 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     ulong lastKnownModifiers = m_lastKnownModifiers;
     ulong delta = lastKnownModifiers ^ modifiers;
     m_lastKnownModifiers = modifiers;
+    const bool swapCtrlAndMeta = !qApp->testAttribute(Qt::AA_MacDontSwapCtrlAndMeta);
 
     struct qt_mac_enum_mapper
     {
@@ -1484,8 +1488,8 @@ static QTabletEvent::TabletDevice wacomTabletDevice(NSEvent *theEvent)
     };
     static qt_mac_enum_mapper modifier_key_symbols[] = {
         { NSShiftKeyMask, Qt::Key_Shift },
-        { NSControlKeyMask, Qt::Key_Meta },
-        { NSCommandKeyMask, Qt::Key_Control },
+        { NSControlKeyMask, (swapCtrlAndMeta ? Qt::Key_Meta :  Qt::Key_Control) },
+        { NSCommandKeyMask, (swapCtrlAndMeta ? Qt::Key_Control : Qt::Key_Meta) },
         { NSAlternateKeyMask, Qt::Key_Alt },
         { NSAlphaShiftKeyMask, Qt::Key_CapsLock },
         { 0ul, Qt::Key_unknown } };
-- 
2.6.3

